### DISCUSSION TITLE
How to "follow" a file (`tail -f`)

### DISCUSSION THREAD

**Discussion by Elias-Graf** [nushell/nushell#7133] [Q&A]
How would I open a log and follow the output of it? On Unix systems, I would usually use `tail -f`. On Windows / PowerShell, `Get-Content -Wait`.

I've found https://github.com/nushell/nushell/discussions/4026, but `last` does not follow by default, and the help doesn't suggest there are any flags to do this.

**Comment by fdncred** [comment]
We do allow opening infinite streams, which you question kind of assumes, I think. However, I haven't personally tried it. My best guess would be something along these lines.
```
open -r somefile.log | lines | each {|f| print $f }
```

**Reply by Elias-Graf** [reply]
Wouldn't it be nice to have a `-t` / `--tail` or `-f`/ `--follow` flag on `last`? Or directly have a `tail` / `follow` command? Seems like something that is used quite a lot.

In addition, those commands I mentioned in the original question, all feature additional flags to initially only print the last x lines. That is really useful when for example trying to follow a large log.

Because I fail to concretely explain what I mean (thus you are assuming), I thought I'd like the documentation of the Unix tail command: https://man7.org/linux/man-pages/man1/tail.1.html :). Please take a look at the description of the `-f` and `-n` flag.

**Reply by Elias-Graf** [reply]
Oh, so I just tried what you mentioned. Seems like I've failed to explain harder than initially thought, sorry 😅. As far as I understand, what you provided, allows me to operate on an infinite amount of lines, correct? What I want, is that the file "is keep open", and newly added lines will be printed as they are added.

What you provides just prints all lines and then exits.

**Reply by fdncred** [reply]
Right. I am familiar with `tail -f`. I've used it quite a bit. I wasn't sure if open would keep it open or not. I remember seeing JT test some infinite stuff and I thought it was with open. 

I'm wondering if @rgwood's `watch` could come in handy here? I haven't tried that either.

**Reply by rgwood** [reply]
You miiiiight be able to hack something together with `watch` but it would be ugly. `watch` is able to tell you when a file was written to, but it doesn't tell you where in the file the write occurred.

**Reply by Elias-Graf** [reply]
Up until now I've only been a user of nu(shell) and never did any scripting in it. But I had some time on my hands and crafted this horrible monster:

```nu
export def tail [file_path: path] {
	watch $file_path { |op|
		if ($op == "Write") {
			print --no-newline (open $file_path | last 1 | get 0 -i)
		}
	}
}
```

It does, very efficiently, open the file on every change :D. I would still appreciate if an actual tail commands with the appropriate flags and optimizations would become standard at some point :)

# EDIT

THIS DOES NOT WORK CORRECTLY!!

For some reason, the watch thingy is not triggered reliably. So, for example, the output is only printed if the application that writes to the log is closed. Funnily enough, it works if I simultaneously have a ` Get-Content -Wait -Tail` on the same file.

# Edit 2

Kind of ridiculous, but hey, better than switching to `pwsh` every time I want to use the functionality :D

```nu
export def tail [file_path: path] {
	pwsh -Command Get-Content -Wait $file_path
}
```

In case anyone wants to look into this:
- How `pwsh` does it: https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/GetContentCommand.cs
- How Unix does it, a bit messy IMO: https://git.savannah.gnu.org/cgit/coreutils.git/tree/src/tail.c

**Comment by bminer** [comment]
It seems clear from this discussion that nu does not support following like `tail` does, so I opened a feature request: https://github.com/nushell/nushell/issues/14529
